store/src/java/com/zimbra/cs/service/admin/HabOrgUnit.java

```diff
@@ -0,0 +1,144 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <https://www.gnu.org/licenses/>.
+ * ***** END LICENSE BLOCK *****
+ */
+package com.zimbra.cs.service.admin;
+
+import java.util.Map;
+import java.util.Set;
+
+import com.zimbra.common.account.Key;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.soap.AdminConstants;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.util.StringUtil;
+import com.zimbra.cs.account.AccountServiceException;
+import com.zimbra.cs.account.Domain;
+import com.zimbra.cs.account.Provisioning;
+import com.zimbra.cs.account.accesscontrol.AdminRight;
+import com.zimbra.cs.ldap.LdapClient;
+import com.zimbra.cs.ldap.LdapServerType;
+import com.zimbra.cs.ldap.LdapUsage;
+import com.zimbra.cs.ldap.ZLdapContext;
+import com.zimbra.soap.ZimbraSoapContext;
+
+
+/**
+ * @author zimbra
+ *
+ */
+public class HabOrgUnit extends AdminDocumentHandler {
+    
+
+    /* (non-Javadoc)
+     * @<HabOrgUnit op=“create/delete/rename” name=“”>
+     * <domain by=“name/id”></domain>
+     * </HabOrgUnit>
+     */
+    @Override
+    public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+        ZimbraSoapContext zsc = getZimbraSoapContext(context);
+        Provisioning prov = Provisioning.getInstance();
+
+        Element d = request.getElement(AdminConstants.E_DOMAIN);
+        String domainKey = d.getAttribute(AdminConstants.A_BY);
+        String domainValue = d.getText();
+        
+        Domain domain = prov.get(Key.DomainBy.fromString(domainKey), domainValue);
+        if (domain == null) {
+            throw AccountServiceException.NO_SUCH_DOMAIN(domainValue);
+        }
+        checkDomainRight(zsc, domain, AdminRight.PR_ALWAYS_ALLOW);
+        if (!domain.isLocal()) {
+            throw ServiceException.INVALID_REQUEST("domain type must be local", null);
+        }
+
+        String habOrgUnitName = request.getAttribute(AdminConstants.A_NAME);
+        if (StringUtil.isNullOrEmpty(habOrgUnitName) ) {
+            throw ServiceException.INVALID_REQUEST(
+                String.format("Hab Org unit name is required, requested name:%s",habOrgUnitName), null);
+        }
+        
+        String operation = request.getAttribute(AdminConstants.A_OPERATION);
+        if (StringUtil.isNullOrEmpty(operation) ) {
+            throw ServiceException.INVALID_REQUEST(
+                String.format("Operation is required, requested name:%s",operation), null);
+        }
+        
+        ZLdapContext zlc = null;
+        Element response = zsc.createElement(AdminConstants.HAB_ORG_UNIT_RESPONSE);
+        
+        try {
+            zlc = LdapClient.getContext(LdapServerType.MASTER, LdapUsage.CREATE_OU);
```

Comment:
Do you need to do this, and the close in `finally` since it's handled in provisioning?

Reply:
Fixed